File: lib/python-sdk/common_grants/schemas/fields.py
"""Base field types and common models for the CommonGrants API."""

from datetime import date, datetime, time
from enum import StrEnum
import re
from typing import Annotated, Any, Optional
from uuid import UUID

from pydantic import Field, HttpUrl, BeforeValidator

from common_grants.schemas.base import CommonGrantsBaseModel

# Date and Time
ISODate = date
ISOTime = time
UTCDateTime = datetime

# DecimalString
def validate_decimal_string(v: str) -> str:
    """Validate a string represents a valid decimal number.
    
    Args:
        v: The string to validate
        
    Returns:
        The validated string
        
    Raises:
        ValueError: If the string is not a valid decimal number
    """
    if not isinstance(v, str):
        raise ValueError("Value must be a string")
    
    if not re.match(r'^-?\d*\.?\d+$', v):
        raise ValueError("Value must be a valid decimal number (e.g., '123.45', '-123.45', '123', '-123')")
    
    return v

DecimalString = Annotated[
    str,
    BeforeValidator(validate_decimal_string),
]

# Money
class Money(CommonGrantsBaseModel):
    """Represents a monetary amount in a specific currency."""

    amount: DecimalString = Field(
        ...,
        description="The amount of money",
        examples=["1000000", "500.00", "-100.50"],
    )
    currency: str = Field(
        ...,
        description="The ISO 4217 currency code (e.g., 'USD', 'EUR')",
        pattern=r"^[A-Z]{3}$",
        min_length=3,
        max_length=3,
        examples=["USD", "EUR", "GBP", "JPY"]
    )


# Event
class Event(CommonGrantsBaseModel):
    """Represents a scheduled event with an optional time and description."""

    name: str = Field(
        ...,
        description="Human-readable name of the event",
        min_length=1,
    )
    date: ISODate = Field(
        ...,
        description="Date of the event in ISO 8601 format: YYYY-MM-DD",
    )
    time: Optional[ISOTime] = Field(
        default=None,
        description="Time of the event in ISO 8601 format: HH:MM:SS",
    )
    description: Optional[str] = Field(
        default=None,
        description="Description of what this event represents",
    )


# CustomField
class CustomFieldType(StrEnum):
    """The type of the custom field."""

    STRING = "string"
    NUMBER = "number"
    BOOLEAN = "boolean"
    OBJECT = "object"
    ARRAY = "array"


class CustomField(CommonGrantsBaseModel):
    """Represents a custom field with type information and validation schema."""

    name: str = Field(
        ...,
        description="Name of the custom field",
        min_length=1,
    )
    type: CustomFieldType = Field(
        ...,
        description="The JSON schema type to use when de-serializing the `value` field",
    )
    schema_url: Optional[HttpUrl] = Field(
        None,
        alias="schema",
        description="Link to the full JSON schema for this custom field",
    )
    value: Any = Field(..., description="Value of the custom field")
    description: Optional[str] = Field(
        None,
        description="Description of the custom field's purpose",
    ) 


# SystemMetadata
class SystemMetadata(CommonGrantsBaseModel):
    """System-managed metadata fields for tracking record creation and modification."""

    created_at: UTCDateTime = Field(
        ...,
        description="The timestamp (in UTC) at which the record was created.",
    )
    last_modified_at: UTCDateTime = Field(
        ...,
        description="The timestamp (in UTC) at which the record was last modified.",
    )

File: lib/ts-sdk/src/schemas/zod/fields.ts
/**
 * Zod schemas for the CommonGrants.Fields namespace in the @common-grants/core library
 *
 * @packageDocumentation
 */

import { z } from "zod";
import { UTCDateTimeSchema, ISODateSchema, ISOTimeSchema, DecimalStringSchema } from "./types";

// ############################################################################
// Event schemas
// ############################################################################

// Event type enum
export const EventTypeEnum = z.enum(["singleDate", "dateRange", "other"]);

// Base event schema (common fields)
const EventBaseSchema = z.object({
  /** Human-readable name of the event (e.g., 'Application posted', 'Question deadline') */
  name: z.string(),

  /** Type of event */
  eventType: EventTypeEnum,

  /** Description of what this event represents */
  description: z.string().nullish(),
});

// Single date event schema
export const SingleDateEventSchema = EventBaseSchema.extend({
  /** Type of event */
  eventType: z.literal("singleDate"),

  /** Date of the event in ISO 8601 format: YYYY-MM-DD */
  date: ISODateSchema,

  /** Time of the event in ISO 8601 format: HH:MM:SS */
  time: ISOTimeSchema.nullish(),
});

// Date range event schema
export const DateRangeEventSchema = EventBaseSchema.extend({
  /** Type of event */
  eventType: z.literal("dateRange"),

  /** Start date of the event in ISO 8601 format: YYYY-MM-DD */
  startDate: ISODateSchema,

  /** Start time of the event in ISO 8601 format: HH:MM:SS */
  startTime: ISOTimeSchema.nullish(),

  /** End date of the event in ISO 8601 format: YYYY-MM-DD */
  endDate: ISODateSchema,

  /** End time of the event in ISO 8601 format: HH:MM:SS */
  endTime: ISOTimeSchema.nullish(),
});

// Other event schema
export const OtherEventSchema = EventBaseSchema.extend({
  /** Type of event */
  eventType: z.literal("other"),

  /** Details of the event's timeline (e.g. "Every other Tuesday") */
  details: z.string().nullish(),

  /** Description of the event */
  description: z.string().nullish(),
});

// Discriminated union for all event types
export const EventSchema = z.discriminatedUnion("eventType", [
  SingleDateEventSchema,
  DateRangeEventSchema,
  OtherEventSchema,
]);

// ############################################################################
// Money schema
// ############################################################################

export const MoneySchema = z.object({
  /** The amount of money */
  amount: DecimalStringSchema,

  /** The ISO 4217 currency code */
  currency: z.string(),
});

// ############################################################################
// CustomField schema
// ############################################################################

export const CustomFieldTypeEnum = z.enum([
  "string",
  "number",
  "integer",
  "boolean",
  "object",
  "array",
]);

export const CustomFieldSchema = z.object({
  /** Name of the custom field */
  name: z.string(),

  /** The JSON schema type */
  fieldType: CustomFieldTypeEnum,

  /** Link to the full JSON schema */
  schema: z.string().url().nullish(),

  /** Value of the custom field */
  value: z.unknown(),

  /** Description of the custom field's purpose */
  description: z.string().nullish(),
});

// ############################################################################
// SystemMetadata schema
// ############################################################################

export const SystemMetadataSchema = z.object({
  /** The timestamp (in UTC) at which the record was created */
  createdAt: UTCDateTimeSchema,

  /** The timestamp (in UTC) at which the record was last modified */
  lastModifiedAt: UTCDateTimeSchema,
});
